### ISSUE TITLE
map_fn + @tf.function + tf.nn.conv2d throws error when strides to conv2d are supplied in the elems argument for map_fn

### ISSUE THREAD

**Issue by aligirayhanozbay** [tensorflow/tensorflow#37460]
<em>Please make sure that this is a bug. As per our
[GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md),
we only address code/doc bugs, performance issues, feature requests and
build/installation issues on GitHub. tag:bug_template</em>

**System information** 
- Have I written custom code (as opposed to using a stock
example script provided in TensorFlow): Yes
- OS Platform and Distribution (e.g.,
Linux Ubuntu 16.04): Ubuntu GNU/Linux 18.04, Debian GNU/Linux bullseye/sid x86_64

- TensorFlow installed from (source or
binary): Binary
- TensorFlow version (use command below): 2.0.0, 2.1.0
- Python version: 3.7.3, 3.5.2
- CUDA/cuDNN version: 10.0, ROCm 3.0
- GPU model and memory: Titan V, Radeon VII

**Describe the current behavior**
Supplying the strides for tf.nn.conv2d inside elems when using tf.map_fn leads to the following error ONLY when executing in graph mode/inside @tf.function:

TypeError: len is not well defined for symbolic Tensors. (TensorArrayV2Read_2/TensorListGetItem:0) Please call `x.shape` rather than `len(x)` for shape information.

**Describe the expected behavior**
It should run successfully, as is the case without @tf.function

**Standalone code to reproduce the issue** 
```python
import tensorflow as tf

@tf.function
def g(a,b):
    return tf.map_fn(lambda x: tf.nn.conv2d(tf.expand_dims(x[0],0),x[1],[2,2],"VALID","NCHW"), [a,b], dtype = a.dtype, parallel_iterations = 16)

def g2(a,b,s):
    return tf.map_fn(lambda x: tf.nn.conv2d(tf.expand_dims(x[0],0),x[1],x[2],"VALID","NCHW"), [a,b,s], dtype = a.dtype, parallel_iterations = 16)

@tf.function
def g3(a,b,s):
    return tf.map_fn(lambda x: tf.nn.conv2d(tf.expand_dims(x[0],0),x[1],x[2],"VALID","NCHW"), [a,b,s], dtype = a.dtype, parallel_iterations = 16)

q = tf.random.uniform((16,5,100,100))
r = tf.random.uniform((16,5,5,5,2))
stri = tf.tile(tf.constant([[2,2]]),[16,1])

print(g(q,r).shape) #runs fine
print(g2(q,r,stri).shape) #runs fine
print(g3(q,r,stri).shape) #error
```

**Other info / logs** Include any logs or source code that would be helpful to
### When using TF2.0 + CUDA 10.0 on Titan V
```
Traceback (most recent call last):
  File "mapfn.py", line 19, in <module>
    print(g3(q,r,stri))
  File "/home/ago14/.conda/envs/tf2.0/lib/python3.7/site-packages/tensorflow_core/python/eager/def_function.py", line 457, in __call__
    result = self._call(*args, **kwds)
  File "/home/ago14/.conda/envs/tf2.0/lib/python3.7/site-packages/tensorflow_core/python/eager/def_function.py", line 503, in _call
    self._initialize(args, kwds, add_initializers_to=initializer_map)
  File "/home/ago14/.conda/envs/tf2.0/lib/python3.7/site-packages/tensorflow_core/python/eager/def_function.py", line 408, in _initialize
    *args, **kwds))
  File "/home/ago14/.conda/envs/tf2.0/lib/python3.7/site-packages/tensorflow_core/python/eager/function.py", line 1848, in _get_concrete_function_internal_garbage_collected
    graph_function, _, _ = self._maybe_define_function(args, kwargs)
  File "/home/ago14/.conda/envs/tf2.0/lib/python3.7/site-packages/tensorflow_core/python/eager/function.py", line 2150, in _maybe_define_function
    graph_function = self._create_graph_function(args, kwargs)
  File "/home/ago14/.conda/envs/tf2.0/lib/python3.7/site-packages/tensorflow_core/python/eager/function.py", line 2041, in _create_graph_function
    capture_by_value=self._capture_by_value),
  File "/home/ago14/.conda/envs/tf2.0/lib/python3.7/site-packages/tensorflow_core/python/framework/func_graph.py", line 915, in func_graph_from_py_func
    func_outputs = python_func(*func_args, **func_kwargs)
  File "/home/ago14/.conda/envs/tf2.0/lib/python3.7/site-packages/tensorflow_core/python/eager/def_function.py", line 358, in wrapped_fn
    return weak_wrapped_fn().__wrapped__(*args, **kwds)
  File "/home/ago14/.conda/envs/tf2.0/lib/python3.7/site-packages/tensorflow_core/python/framework/func_graph.py", line 905, in wrapper
    raise e.ag_error_metadata.to_exception(e)
TypeError: in converted code:

    mapfn.py:12 g3  *
        return tf.map_fn(lambda x: tf.nn.conv2d(tf.expand_dims(x[0],0),x[1],x[2],"VALID","NCHW"), [a,b,s], dtype = a.dtype, parallel_iterations = 16)
    /home/ago14/.conda/envs/tf2.0/lib/python3.7/site-packages/tensorflow_core/python/ops/map_fn.py:268 map_fn
        maximum_iterations=n)
    /home/ago14/.conda/envs/tf2.0/lib/python3.7/site-packages/tensorflow_core/python/ops/nn_ops.py:1913 conv2d_v2
        name=name)
    /home/ago14/.conda/envs/tf2.0/lib/python3.7/site-packages/tensorflow_core/python/ops/nn_ops.py:2000 conv2d
        strides = _get_sequence(strides, 2, channel_index, "strides")
    /home/ago14/.conda/envs/tf2.0/lib/python3.7/site-packages/tensorflow_core/python/ops/nn_ops.py:68 _get_sequence
        current_n = len(value)
    /home/ago14/.conda/envs/tf2.0/lib/python3.7/site-packages/tensorflow_core/python/framework/ops.py:741 __len__
        "shape information.".format(self.name))

    TypeError: len is not well defined for symbolic Tensors. (TensorArrayV2Read_2/TensorListGetItem:0) Please call `x.shape` rather than `len(x)` for shape information.
```

### When using TF2.1 + ROCm 3.0 on Radeon VII
```
Traceback (most recent call last):
  File "mapfn.py", line 20, in <module>
    print(g3(q,r,stri).shape) #error
  File "/usr/local/lib/python3.5/dist-packages/tensorflow_core/python/eager/def_function.py", line 568, in __call__
    result = self._call(*args, **kwds)
  File "/usr/local/lib/python3.5/dist-packages/tensorflow_core/python/eager/def_function.py", line 615, in _call
    self._initialize(args, kwds, add_initializers_to=initializers)
  File "/usr/local/lib/python3.5/dist-packages/tensorflow_core/python/eager/def_function.py", line 497, in _initialize
    *args, **kwds))
  File "/usr/local/lib/python3.5/dist-packages/tensorflow_core/python/eager/function.py", line 2389, in _get_concrete_function_internal_garbage_collected
    graph_function, _, _ = self._maybe_define_function(args, kwargs)
  File "/usr/local/lib/python3.5/dist-packages/tensorflow_core/python/eager/function.py", line 2703, in _maybe_define_function
    graph_function = self._create_graph_function(args, kwargs)
  File "/usr/local/lib/python3.5/dist-packages/tensorflow_core/python/eager/function.py", line 2593, in _create_graph_function
    capture_by_value=self._capture_by_value),
  File "/usr/local/lib/python3.5/dist-packages/tensorflow_core/python/framework/func_graph.py", line 978, in func_graph_from_py_func
    func_outputs = python_func(*func_args, **func_kwargs)
  File "/usr/local/lib/python3.5/dist-packages/tensorflow_core/python/eager/def_function.py", line 439, in wrapped_fn
    return weak_wrapped_fn().__wrapped__(*args, **kwds)
  File "/usr/local/lib/python3.5/dist-packages/tensorflow_core/python/framework/func_graph.py", line 968, in wrapper
    raise e.ag_error_metadata.to_exception(e)
TypeError: in converted code:

    mapfn.py:12 g3  *
        return tf.map_fn(lambda x: tf.nn.conv2d(tf.expand_dims(x[0],0),x[1],x[2],"VALID","NCHW"), [a,b,s], dtype = a.dtype, parallel_iterations = 16)
    /usr/local/lib/python3.5/dist-packages/tensorflow_core/python/ops/map_fn.py:268 map_fn
        maximum_iterations=n)
    /usr/local/lib/python3.5/dist-packages/tensorflow_core/python/ops/nn_ops.py:1914 conv2d_v2
        name=name)
    /usr/local/lib/python3.5/dist-packages/tensorflow_core/python/ops/nn_ops.py:2001 conv2d
        strides = _get_sequence(strides, 2, channel_index, "strides")
    /usr/local/lib/python3.5/dist-packages/tensorflow_core/python/ops/nn_ops.py:69 _get_sequence
        current_n = len(value)
    /usr/local/lib/python3.5/dist-packages/tensorflow_core/python/framework/ops.py:733 __len__
        "shape information.".format(self.name))

    TypeError: len is not well defined for symbolic Tensors. (TensorArrayV2Read_2/TensorListGetItem:0) Please call `x.shape` rather than `len(x)` for shape information.
```

**Comment by Saduf2019** [comment]
@aligirayhanozbay 
PLease refer to this [link](https://github.com/wau/keras-rl2/pull/9), [link2](https://github.com/keras-rl/keras-rl/issues/348) and [link3](https://stackoverflow.com/questions/59682542/typeerror-len-is-not-well-defined-for-symbolic-tensors-activation-3-identity), let us know if this helps resolve your issue.

**Comment by aligirayhanozbay** [comment]
Dear @Saduf2019 ,
I am not using keras-rl, so I believe this issue is not directly related to the issues you linked. May I ask why a len() check is being done on the strides argument of tf.nn.conv2d only?

**Comment by Saduf2019** [comment]
@aligirayhanozbay 
Could you please check [this link](https://stackoverflow.com/questions/34642595/tensorflow-strides-argument) in reference to len().

Also This question is better asked on [StackOverflow](http://stackoverflow.com/questions/tagged//tensorflow) since it is not a bug or feature request. There is also a larger community that reads questions there.

**Comment by Saduf2019** [comment]
@aligirayhanozbay
please update on the above comment

**Comment by aligirayhanozbay** [comment]
Dear @Saduf2019,

Sorry for the belated reply - I did not have access to the machines I was testing this on for a little while. I read the stackoverflow link you posted, but I disagree that there is no bug involved here. It seems like tf.nn.conv2d has been written to accept a list (and not a tensor) for the strides argument, but map_fn cannot supply a list to conv2d. The number of elements supplied to conv2d does not make a difference - I get precisely the same errors if I replace the line `stri = tf.tile(tf.constant([[2,2]]),[16,1])` with `stri = tf.tile(tf.constant([[1,1,2,2]]),[16,1])`.

Supplying a list causes different errors since map_fn is not capable of dealing with them properly. Running this snippet of code
```python
@tf.function
def g3(a,b,s):
    return tf.map_fn(lambda x: tf.nn.conv2d(tf.expand_dims(x[0],0),x[1],x[2],"VALID","NCHW"), [a,b,s], dtype = a.dtype, parallel_iterations = 16)

q = tf.random.uniform((16,5,100,100))
r = tf.random.uniform((16,5,5,5,2))

stri = [[1,1,2,2] for _ in range(16)]

print(g3(q,r,stri).shape) #error
```
leads to the following traceback
```
Traceback (most recent call last):
  File "mapfn.py", line 23, in <module>
    print(g3(q,r,stri).shape) #error
  File "/usr/local/lib/python3.5/dist-packages/tensorflow_core/python/eager/def_function.py", line 568, in __call__
    result = self._call(*args, **kwds)
  File "/usr/local/lib/python3.5/dist-packages/tensorflow_core/python/eager/def_function.py", line 615, in _call
    self._initialize(args, kwds, add_initializers_to=initializers)
  File "/usr/local/lib/python3.5/dist-packages/tensorflow_core/python/eager/def_function.py", line 497, in _initialize
    *args, **kwds))
  File "/usr/local/lib/python3.5/dist-packages/tensorflow_core/python/eager/function.py", line 2389, in _get_concrete_function_internal_garbage_collected
    graph_function, _, _ = self._maybe_define_function(args, kwargs)
  File "/usr/local/lib/python3.5/dist-packages/tensorflow_core/python/eager/function.py", line 2703, in _maybe_define_function
    graph_function = self._create_graph_function(args, kwargs)
  File "/usr/local/lib/python3.5/dist-packages/tensorflow_core/python/eager/function.py", line 2593, in _create_graph_function
    capture_by_value=self._capture_by_value),
  File "/usr/local/lib/python3.5/dist-packages/tensorflow_core/python/framework/func_graph.py", line 978, in func_graph_from_py_func
    func_outputs = python_func(*func_args, **func_kwargs)
  File "/usr/local/lib/python3.5/dist-packages/tensorflow_core/python/eager/def_function.py", line 439, in wrapped_fn
    return weak_wrapped_fn().__wrapped__(*args, **kwds)
  File "/usr/local/lib/python3.5/dist-packages/tensorflow_core/python/framework/func_graph.py", line 968, in wrapper
    raise e.ag_error_metadata.to_exception(e)
ValueError: in converted code:

    mapfn.py:13 g3  *
        return tf.map_fn(lambda x: tf.nn.conv2d(tf.expand_dims(x[0],0),x[1],x[2],"VALID","NCHW"), [a,b,s], dtype = a.dtype, parallel_iterations = 16)
    /usr/local/lib/python3.5/dist-packages/tensorflow_core/python/ops/map_fn.py:268 map_fn
        maximum_iterations=n)
    /usr/local/lib/python3.5/dist-packages/tensorflow_core/python/ops/nn_ops.py:1914 conv2d_v2
        name=name)
    /usr/local/lib/python3.5/dist-packages/tensorflow_core/python/ops/nn_ops.py:2001 conv2d
        strides = _get_sequence(strides, 2, channel_index, "strides")
    /usr/local/lib/python3.5/dist-packages/tensorflow_core/python/ops/nn_ops.py:78 _get_sequence
        name, n, n + 2, current_n))

    ValueError: strides should be of length 1, 2 or 4 but was 16

```
which seems to imply to me that it tried to pass the whole list into each slice map_fn is operating on. A variation with a list of 16 integers supplied leads to the same error. However, things are not so straightforward since running the above snippet with `stri=[1,1,2,2]` leads to 
```
mapfn.py:13 g3  *
        return tf.map_fn(lambda x: tf.nn.conv2d(tf.expand_dims(x[0],0),x[1],x[2],"VALID","NCHW"), [a,b,s], dtype = a.dtype, parallel_iterations = 16)
    /usr/local/lib/python3.5/dist-packages/tensorflow_core/python/ops/map_fn.py:268 map_fn
        maximum_iterations=n)
    /usr/local/lib/python3.5/dist-packages/tensorflow_core/python/ops/nn_ops.py:1914 conv2d_v2
        name=name)
    /usr/local/lib/python3.5/dist-packages/tensorflow_core/python/ops/nn_ops.py:2011 conv2d
        name=name)
    /usr/local/lib/python3.5/dist-packages/tensorflow_core/python/ops/gen_nn_ops.py:943 conv2d
        strides = [_execute.make_int(_i, "strides") for _i in strides]
    /usr/local/lib/python3.5/dist-packages/tensorflow_core/python/ops/gen_nn_ops.py:943 <listcomp>
        strides = [_execute.make_int(_i, "strides") for _i in strides]
    /usr/local/lib/python3.5/dist-packages/tensorflow_core/python/eager/execute.py:174 make_int
        (arg_name, repr(v)))

    TypeError: Expected int for argument 'strides' not <tf.Tensor 'TensorArrayV2Read_2/TensorListGetItem:0' shape=() dtype=int32>.
```
Just for the sake of completeness, I'd like to emphasize that supplying any sort of tensor (e.g. tf.constant([2 for _ in range(16)])) also lead to a traceback similar to one in the original post. 

To recap, it seems that tf.nn.conv2d expects to see a list (at least in graph mode, oddly enough) to specify the strides argument. However, map_fn is unable to supply this argument as a list (supplying [[1,1,2,2] for _ in range(16)] does not end up in passing [1,1,2,2] to each slice map_fn is operating on) which leads to an error. It is also impossible to call list() on the argument inside tf.function.

**Comment by Saduf2019** [comment]
i have replicated the code shared, please find the [gist here](https://colab.sandbox.google.com/gist/Saduf2019/a23c29bfb3df4256d881f54ed7fbe66a/37460.ipynb#scrollTo=xqR-X3yVQ6wo)

**Comment by aligirayhanozbay** [comment]
Looks like the nightly version returns a different error. Any reason why that may be?

Also, the original error is still there in the nightly version, if you try to instead run 
```python
@tf.function
def g3(a,b,s):
    return tf.map_fn(lambda x: tf.nn.conv2d(tf.expand_dims(x[0],0),x[1],x[2],"VALID","NHWC"), [a,b,s], dtype = a.dtype, parallel_iterations = 16)

q = tf.random.uniform((16,100,100,5))
r = tf.random.uniform((16,5,5,5,2))

#stri = [[1,1,2,2] for _ in range(16)]
stri = tf.tile(tf.constant([[2,2]]),[16,1])

print(g3(q,r,stri).shape) #error
```

but commenting out the @tf.function, we get the intended result:
```python
#@tf.function
def g3(a,b,s):
    return tf.map_fn(lambda x: tf.nn.conv2d(tf.expand_dims(x[0],0),x[1],x[2],"VALID","NHWC"), [a,b,s], dtype = a.dtype, parallel_iterations = 16)

q = tf.random.uniform((16,100,100,5))
r = tf.random.uniform((16,5,5,5,2))

#stri = [[1,1,2,2] for _ in range(16)]
stri = tf.tile(tf.constant([[2,2]]),[16,1])

print(g3(q,r,stri).shape) #error
```
with the correct shape `(16, 1, 48, 48, 2)` printed.

(Please note that I changed the tensor format to NHWC and, correspondingly, the shape of the tensor q because colab seems to run on CPU, which for some reason cannot do NCHW)

**Comment by sachinprasadhs** [comment]
Was able to reproduce your issue in Tf Nightly 2.6.0-dev20210524, please find the gist [here](https://colab.research.google.com/gist/sachinprasadhs/19ca969122707b95b89ba9a25bac6c85/untitled7.ipynb). Thanks!

**Comment by rohan100jain** [comment]
The underlying problem here is that the strides argument for `tf.nn.conv2d` cannot be a Tensor (https://www.tensorflow.org/api_docs/python/tf/nn/conv2d). It needs to be a static int or list of ints. In Eager mode (without tf.function), a Tensor input is as good as something static as we know its value but in tf.function we don't support that.

If you want tf.function, then I'm afraid the only solution here is to use g() where the strides argument is hard coded. Or you can do something like

```python
@tf.function
def g4(a, b, s):
  return tf.map_fn(lambda x: tf.nn.conv2d(tf.expand_dims(x[0], 0), x[1], s, "VALID", "NCHW"), [a, b], dtype=a.dtype, parallel_iterations=16)

q = tf.random.uniform((16,100,100,5))
r = tf.random.uniform((16,5,5,5,2))
stri = [2, 2]

print(g4(q,r,stri).shape) 
```

Everytime you change stri to something else, we'll most likely recompile the tf.function as it'll change the function cache.

**Comment by aligirayhanozbay** [comment]
@rohan100jain I solved my issue over the past year with a similar approach to the snippet in your comment; please feel free to close the issue if the behaviour of `tf.nn.conv2d` won't be changed in graph mode to accommodate tensors as the stride argument.
